apps/ejabberd/src/ejabberd_s2s_out.erl

```diff
@@ -467,42 +450,18 @@ wait_for_features({xmlstreamelement, El}, StateData) ->
             {SASLEXT, StartTLS, StartTLSRequired} =
                 lists:foldl(
                   fun(#xmlel{name = <<"mechanisms">>, attrs = Attrs1,
-                             children = Els1} = _El1,
-                      {_SEXT, STLS, STLSReq} = Acc) ->
-                          case xml:get_attr_s(<<"xmlns">>, Attrs1) of
-                              ?NS_SASL ->
-                                  NewSEXT =
-                                      lists:any(
-                                        fun(#xmlel{name = <<"mechanism">>,
-                                                   children = Els2}) ->
-                                                case xml:get_cdata(Els2) of
-                                                    <<"EXTERNAL">> -> true;
-                                                    _ -> false
-                                                end;
-                                           (_) -> false
-                                        end, Els1),
-                                  {NewSEXT, STLS, STLSReq};
-                              _ ->
-                                  Acc
-                          end;
-                     (#xmlel{name = <<"starttls">>, attrs = Attrs1} = El1,
-                      {SEXT, _STLS, _STLSReq} = Acc) ->
-                          case xml:get_attr_s(<<"xmlns">>, Attrs1) of
-                              ?NS_TLS ->
-                                  Req = case xml:get_subtag(El1, <<"required">>) of
-                                            #xmlel{} -> true;
-                                            false -> false
-                                        end,
-                                  {SEXT, true, Req};
-                              _ ->
-                                  Acc
-                          end;
+                             children = Els1} = _El1, Acc) ->
+                          Attr =  xml:get_attr_s(<<"xmlns">>, Attrs1),
+                          get_acc_with_new_sext(Attr, Els1, Acc);
+                     (#xmlel{name = <<"starttls">>, attrs = Attrs1} = El1, Acc) ->
+                          Attr = xml:get_attr_s(<<"xmlns">>, Attrs1),
+                          get_acc_with_new_req(Attr, El1, Acc);
                      (_, Acc) ->
                           Acc
                   end, {false, false, false}, Els),
-            if
-                (not SASLEXT) and (not StartTLS) and
-                StateData#state.authenticated ->
+            case {SASLEXT, StartTLS, StartTLSRequired, StateData} of
+                {SASLEXT, StartTLS, _, StateData} when (not SASLEXT) and (not StartTLS) and
+                                                       StateData#state.authenticated ->
```

Comment:
ibidem

Reply:
@bartekgorny What do you mean here?

Reply:
I mean the same as above - turn it into a set of pattern-matching functions, would improve readability.